### PR TITLE
Watch disabled property of  option/optgroup elements

### LINKED ISSUE #3347: Enabled/disabled state not respected on options
http://jsfiddle.net/upgrmrL3/1/

When clicking the disable button all options get disabled. When clicking "enable" the options in the select2 dropdown still appear as disabled.

screen shot 2015-05-12 at 13 53 06

Sometimes the disable attribute is not respected either: by clicking disable the options in the dropdown won't get disabled.

screen-shot-2015-05-12-at-13 54 13

The screenshots were taken on the same fiddle, with some cache clearing and refreshes between them.
Testing was done on Firefox 37.0.2 on Mac OS

### PR DESCRIPTION
**Pull request by letrunghieu**
This pull request includes a
- [x] Bug fix
- [ ] New feature
- [ ] Translation

The following changes were made
- Support `disabled` property on `optgroup` elements
- Watch the `disabled` property on `optgroup` and `option` elements so that changes of this property are respected.

Fix #3347, #2306

Demo: https://jsbin.com/moduda/3/edit?html,js,output

### CODE REVIEW

**Comment by md5** [comment]
It looks like this doesn't cover the case when an already selected option is disabled.

File: src/js/select2/data/select.js [hunk 1]
```diff
@@ -203,6 +203,14 @@ define([
     data = $.data($option[0], 'data');
 
     if (data != null) {
+      // always update the value of `disabled` property of the option element
+      data.disabled = $option.prop('disabled');
```
Comment (kevin-brown) [inline src/js/select2/data/select.js]: Ideally the solution to this should be very similar, if not the same as, the solution to us handling the cached `selected` property on elements.

https://github.com/select2/select2/commit/54154c3df6b7e2f59dc76b663f230f9b773397f2

**Comment by alexweissman** [comment]
@letrunghieu are you still interested in working on this?  I'd like to merge it, but a few things need to be addressed:
1. Can you address @kevin-brown's review comment?
2. Can you write some qunit test cases?
3. Can you remove the changes in `dist/` from the PR?

Thanks!

**Comment by letrunghieu** [comment]
Hi @alexweissman I am sure interested in working on this issue :) I'll review the code.

File: src/js/select2/data/select.js (hunk 1 above)
Comment (denilsonsa) [reply src/js/select2/data/select.js]: We now have a PR that is similar to the mentioned commit: #4818

Comment (alexweissman) [reply src/js/select2/data/select.js]: @denilsonsa I don't think that #4818 handles `optgroup` elements.  You'll notice that @letrunghieu's solution does a bit more.  However, neither of these PRs have any tests yet 😦

**Comment by letrunghieu** [comment]
Hi @kevin-brown , @alexweissman , could you review this PR again please?

**Comment by alexweissman** [comment]
Looking good.  Can we avoid the `!!` method of type coercion?  It's a little bit obscure.

**Comment by letrunghieu** [comment]
@alexweissman So how to we make sure that the type of the `disabled` variable is boolean?

**Comment by alexweissman** [comment]
Is [`Boolean`](https://stackoverflow.com/a/31155542/2970321) widely supported?

**Comment by denilsonsa** [comment]
Should be, as it is part of JavaScript since (basically) forever: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean

(Hi! I haven't contributed to this project yet; I just wanted to share this relevant link!)

**Pull request closed without merging**